### ISSUE TITLE
Grace Termination Period timer doesn't start before preStop

### ISSUE THREAD

**Issue by cyrus-mc** [kubernetes/kubernetes#117373]
### What happened?

Implementing graceful shutdown by leveraging `preStop` hooks. According to all the documentation I can find the `terminatinoGracePeriodSeconds` includes the time that the `preStop` is running, meaning the timer starts before `preStop` executes.

In my setup I have a simple `preStop` that sleeps for 15 seconds. In my app code I then sleep for 90 seconds in the SIGTERM handler. Given that the default terminationGracePeriod is 30 seconds I would expect that the POD is killed 15 seconds after the SIGTERM handler reports it received the signal. However the POD receives the SIGTERM (prints out that it did) and then lives for a while 30 seconds after that point. 

### What did you expect to happen?

As it states here:

* https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/
* https://www.datree.io/resources/kubernetes-guide-graceful-shutdown-with-lifecycle-prestop-hook#anchor5
* https://livebook.manning.com/concept/kubernetes/terminationgraceperiodsecond

```
The Pod's termination grace period countdown begins before the PreStop hook is executed, so regardless of the outcome of the handler, the container will eventually terminate within the Pod's termination grace period.
```

### How can we reproduce it (as minimally and precisely as possible)?

I have the following simple NodeJS app:

```
import * as Koa from 'koa';
import * as Router from 'koa-router';

const app = new Koa();
const router = new Router();

function delay(time) {
  return new Promise(resolve => setTimeout(resolve, time));
}

router.get("/", ctx => {
  ctx.body = 'Hello World';
});

app.use(router.routes);
app.use(router.allowedMethods());

export const server = app.listen(3000, () => {
  console.log(`Server listening on port 3000`);
});

process.on('SIGTERM', () => {
  server.close(async() => {
    console.log('Received SIGTERM, closing server');
    await delay(90000);
  });
});
```

As you can see in the SIGTERM handler it prints out the signal was received and then simply sleeps for 90 seconds.

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: debug
  namespace: default
  labels:
    app: debug
spec:
  replicas: 1
  selector:
    matchLabels:
      app: debug
  template:
    metadata:
      labels:
        app: debug
    spec:
      imagePullSecrets:
        - name: dockerconfigjson-jfrog
      containers:
      - name: node
        image: test:latest
        lifecycle:
          preStop:
            exec:
              command:
                - sleep
                - "15"
        resources:
          requests:
            memory: 128Mi
          limits:
            memory: 256Mi
```

My deployment of above code adds a preStop that simply sleeps for 15 seconds. 

When I delete this POD, the application receives the SIGTERM after 15 seconds (as expected). It then lives for a whole 30 more seconds before being forcefully terminated. Give that the `terminationGracePeriodSeconds` is set to 30 seconds if I am interpreting the documentation above correctly the POD should have been forcefully terminated at 15 seconds after receiving SIGTERM.

```
    lifecycle:
      preStop:
        exec:
          command:
          - sleep
          - "15"
    name: node
    resources:
      limits:
        memory: 256Mi
      requests:
        memory: 128Mi
  securityContext:
    seccompProfile:
      type: RuntimeDefault
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30

```

### Kubernetes version

<details>

```console
› kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
WARNING: version difference between client (1.25) and server (1.23) exceeds the supported minor version skew of +/-1
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.4", GitCommit:"872a965c6c6526caa949f0c6ac028ef7aff3fb78", GitTreeState:"clean", BuildDate:"2022-11-09T13:36:36Z", GoVersion:"go1.19.3", Compiler:"gc", Platform:"darwin/arm64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"23+", GitVersion:"v1.23.16-eks-48e63af", GitCommit:"e6332a8a3feb9e0fe3db851878f88cb73d49dd7a", GitTreeState:"clean", BuildDate:"2023-01-24T19:18:15Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

Running in AWS EKS version 1.23.

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

Container Runtime Version:  containerd://1.6.15+bottlerocket

**Comment by smarterclayton** [comment]
> However the POD receives the SIGTERM (prints out that it did) and then lives for a while 30 seconds after that point.

Just verifying, you did:

1. T0: Delete pod with 30s grace period and 15s prestop Hook
2. T0+: Observe preStop run
3. T15~: Observe SIGTERM
4. T30~: Expected SIGKILL
5. T45~: Observed pod stop

?

**Comment by smarterclayton** [comment]
Your interpretation of the intended behavior (preStop time is charged to graceful shutdown) is correct.  The gap is that the CRI implementation did not fully implement the described behavior.

/assign

to me for now because I'm coordinating correct behavior for graceful termination like #83916.

**Comment by cyrus-mc** [comment]
> > However the POD receives the SIGTERM (prints out that it did) and then lives for a while 30 seconds after that point.
> 
> Just verifying, you did:
> 
> 1. T0: Delete pod with 30s grace period and 15s prestop Hook
> 2. T0+: Observe preStop run
> 3. T15~: Observe SIGTERM
> 4. T30~: Expected SIGKILL
> 5. T45~: Observed pod stop
> 
> ?

That is correct

**Comment by smarterclayton** [comment]
Thanks, this needs to be fixed and is high on my priority list.

**Comment by alex-richman-onesignal** [comment]
We're also seeing this issue

Pods have a preStop equivalent to a 300s sleep, but ignoring SIGTERM, and a terminationGracePeriod: 300

But on term the pods take 10 minutes rather than 5 minutes to stop.

Here's some logs from the node which show the issue:
```
### Pod deployment is restarted, delete event fires, pod goes into Terminating, preStop hook is called
Apr 29 15:20:01 kubelet[2384]: "SyncLoop DELETE" source="api" pods=[<redacted>-production/<redacted>-68c6fc9b7d-26fwh]


### 300s later (terminationGracePeriodSeconds) the prestop hook has not completed, SIGTERM is fired, but apparently with a new 300s grace period?
Apr 29 15:25:01 kubelet[2384]: "PreStop hook not completed in grace period" pod="<redacted>-production/<redacted>-68c6fc9b7d-26fwh" podUID=f5d85c9b-ed4e-44d4-a075-7eb277ddcc85 containerName="<redacted>" containerID="containerd://d5be3864132695f5b23fa572b4bc183263db2e3ad4d22f2a8871b734a3c5bc13" gracePeriod=300
Apr 29 15:25:01 kubelet[2384]: "Killing container with a grace period" pod="<redacted>-production/<redacted>-68c6fc9b7d-26fwh" podUID=f5d85c9b-ed4e-44d4-a075-7eb277ddcc85 containerName="<redacted>" containerID="containerd://d5be3864132695f5b23fa572b4bc183263db2e3ad4d22f2a8871b734a3c5bc13" gracePeriod=300
Apr 29 15:25:01 containerd[2102]: time="2023-04-29T15:25:01.378199364Z" level=info msg="StopContainer for \"d5be3864132695f5b23fa572b4bc183263db2e3ad4d22f2a8871b734a3c5bc13\" with timeout 300 (s)"


### 300s later, the pod is SIGKILL'd
Apr 29 15:30:01 containerd[2102]: time="2023-04-29T15:30:01.993119307Z" level=info msg="StopContainer for \"d5be3864132695f5b23fa572b4bc183263db2e3ad4d22f2a8871b734a3c5bc13\" returns successfully"
Apr 29 15:30:02 kubelet[2384]: "SyncLoop REMOVE" source="api" pods=[<redacted>-production/<redacted>-68c6fc9b7d-26fwh]
```


This is running on GKE v1.24.10-gke.2300 (same control plane + node pool version).

**Comment by zcahana** [comment]
I believe this has been addressed by #115835, and can be closed now.

**Comment by HirazawaUi** [comment]
Also think https://github.com/kubernetes/kubernetes/pull/115835 has addressed this issue, if there are any other thoughts please reopen this issue
